src/term/check/type_check.rs

```diff
@@ -0,0 +1,70 @@
+use core::fmt;
+use std::collections::HashMap;
+
+use crate::term::{Book, DefId, Name, RulePat};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+  Any,
+  Adt(Name),
+}
+
+pub type DefinitionTypes = HashMap<DefId, Vec<Type>>;
+
+impl Book {
+  /// Returns a HashMap from the definition id to the inferred pattern types
+  /// and checks the rules arities based on the first rule arity.
+  pub fn type_check(&self) -> anyhow::Result<DefinitionTypes> {
+    let mut def_types = HashMap::new();
+
+    for def in self.defs.values() {
+      let current_arity = def.arity();
```

Comment:
Can you separate this into another function? So we can get the type of just a single definition given the existing type definitions

Reply:
`infer_def_type`